File: app/src/main/java/com/nikola/jakshic/dagger/view/activity/HomeActivity.java
package com.nikola.jakshic.dagger.view.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;

import com.nikola.jakshic.dagger.R;
import com.nikola.jakshic.dagger.view.fragment.BookmarkFragment;
import com.nikola.jakshic.dagger.view.fragment.CompetitiveFragment;
import com.nikola.jakshic.dagger.view.fragment.LeaderboardFragment;

public class HomeActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        setTitle("Title");
        BottomNavigationView btmNavView = findViewById(R.id.btm_navigation_home);

        Fragment competitive;
        Fragment bookmark;
        Fragment leaderboard;

        if (savedInstanceState == null) {
            competitive = new CompetitiveFragment();
            bookmark = new BookmarkFragment();
            leaderboard = new LeaderboardFragment();

            getSupportFragmentManager().beginTransaction()
                    .add(R.id.home_fragment_container, competitive, "competitive-tag")
                    .add(R.id.home_fragment_container, leaderboard, "leaderboard-tag")
                    .add(R.id.home_fragment_container, bookmark, "bookmark-tag")
                    .hide(bookmark)
                    .hide(leaderboard)
                    .commit();
        } else {
            competitive = getSupportFragmentManager().findFragmentByTag("competitive-tag");
            bookmark = getSupportFragmentManager().findFragmentByTag("bookmark-tag");
            leaderboard = getSupportFragmentManager().findFragmentByTag("leaderboard-tag");
        }

        btmNavView.setOnNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_competitive:
                    showFragment(competitive, true);
                    showFragment(leaderboard, false);
                    showFragment(bookmark, false);
                    setTitle("Competitive");
                    return true;
                case R.id.action_leaderboard:
                    showFragment(competitive, false);
                    showFragment(leaderboard, true);
                    showFragment(bookmark, false);
                    setTitle("Leaderboard");
                    return true;
                case R.id.action_bookmark:
                    showFragment(competitive, false);
                    showFragment(leaderboard, false);
                    showFragment(bookmark, true);
                    setTitle("Bookmark");
                    return true;
            }
            return false;
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.menu_home_search:
                startActivity(new Intent(this, SearchActivity.class));
                return true;
            case R.id.menu_home_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void showFragment(Fragment fragment, boolean show) {
        FragmentTransaction transaction = getSupportFragmentManager()
                .beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        if (show)
            transaction.show(fragment);
        else
            transaction.hide(fragment);
        transaction.commit();
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/HomeActivity.kt
package com.nikola.jakshic.dagger.ui

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.AppCompatActivity
import com.nikola.jakshic.dagger.R
import com.nikola.jakshic.dagger.hide
import com.nikola.jakshic.dagger.show
import com.nikola.jakshic.dagger.ui.bookmark.BookmarkFragment
import com.nikola.jakshic.dagger.ui.competitive.CompetitiveFragment
import com.nikola.jakshic.dagger.ui.leaderboard.LeaderboardFragment
import com.nikola.jakshic.dagger.ui.stream.StreamFragment
import kotlinx.android.synthetic.main.activity_home.*

class HomeActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)

        val competitive: Fragment
        val leaderboard: Fragment
        val bookmark: Fragment
        val stream: Fragment

        if (savedInstanceState == null) {
            competitive = CompetitiveFragment()
            leaderboard = LeaderboardFragment()
            bookmark = BookmarkFragment()
            stream = StreamFragment()

            supportFragmentManager.beginTransaction()
                    .add(R.id.home_fragment_container, competitive, "competitive-tag")
                    .add(R.id.home_fragment_container, leaderboard, "leaderboard-tag")
                    .add(R.id.home_fragment_container, bookmark, "bookmark-tag")
                    .add(R.id.home_fragment_container, stream, "stream-tag")
                    .hide(leaderboard)
                    .hide(bookmark)
                    .hide(stream)
                    .commit()
        } else {
            competitive = supportFragmentManager.findFragmentByTag("competitive-tag")!!
            leaderboard = supportFragmentManager.findFragmentByTag("leaderboard-tag")!!
            bookmark = supportFragmentManager.findFragmentByTag("bookmark-tag")!!
            stream = supportFragmentManager.findFragmentByTag("stream-tag")!!
        }

        btmNavigation.setOnNavigationItemSelectedListener {
            when (it.itemId) {
                R.id.action_competitive -> {
                    competitive.show()
                    leaderboard.hide()
                    bookmark.hide()
                    stream.hide()
                    return@setOnNavigationItemSelectedListener true
                }
                R.id.action_leaderboard -> {
                    competitive.hide()
                    leaderboard.show()
                    bookmark.hide()
                    stream.hide()
                    return@setOnNavigationItemSelectedListener true
                }
                R.id.action_bookmark -> {
                    competitive.hide()
                    leaderboard.hide()
                    bookmark.show()
                    stream.hide()
                    return@setOnNavigationItemSelectedListener true
                }
                R.id.action_stream -> {
                    competitive.hide()
                    leaderboard.hide()
                    bookmark.hide()
                    stream.show()
                    return@setOnNavigationItemSelectedListener true
                }
                else -> return@setOnNavigationItemSelectedListener false
            }
        }

        btmNavigation.setOnNavigationItemReselectedListener {
            when (it.itemId) {
                R.id.action_competitive -> (competitive as CompetitiveFragment).onItemReselected()
                R.id.action_leaderboard -> (leaderboard as LeaderboardFragment).onItemReselected()
                R.id.action_bookmark -> (bookmark as BookmarkFragment).onItemReselected()
                R.id.action_stream -> (stream as StreamFragment).onItemReselected()
            }
        }
    }

    override fun onBackPressed() {
        if (btmNavigation.selectedItemId != R.id.action_competitive)
            btmNavigation.selectedItemId = R.id.action_competitive
        else
            super.onBackPressed()
    }

    interface OnNavigationItemReselectedListener {

        fun onItemReselected()
    }
}